### ISSUE TITLE
XML File Patching - SyntaxError: Unexpected number

### ISSUE THREAD

**Issue by zaengi** [geeklearningio/gl-vsts-tasks-file-patch#40]
Using Slick Syntax on your newest toolset version  (Version 2.2.999) leads to an error, which I rather cant debug.

2017-08-23T18:24:12.4598851Z Set workingFolder to default: C:\vsoAgent\tasks\XmlPatch\2.2.999
2017-08-23T18:24:12.9039025Z ##[debug]agent.workFolder=C:\vsoAgent\_work
2017-08-23T18:24:12.9039025Z ##[debug]loading inputs and endpoints
2017-08-23T18:24:12.9049040Z ##[debug]loading ENDPOINT_AUTH_$/
2017-08-23T18:24:12.9059043Z ##[debug]loading ENDPOINT_AUTH_SYSTEMVSSCONNECTION
2017-08-23T18:24:12.9059043Z ##[debug]loading INPUT_FAILIFNOPATCHAPPLIED
2017-08-23T18:24:12.9069047Z ##[debug]loading INPUT_JSONPATCHCONTENT
2017-08-23T18:24:12.9069047Z ##[debug]loading INPUT_OUTPUTPATCHFILE
2017-08-23T18:24:12.9079019Z ##[debug]loading INPUT_SKIPERRORS
2017-08-23T18:24:12.9079019Z ##[debug]loading INPUT_SYNTAXTYPE
2017-08-23T18:24:12.9089036Z ##[debug]loading INPUT_XMLTARGETFILTERS
2017-08-23T18:24:12.9089036Z ##[debug]loading INPUT_XMLWORKINGDIR
2017-08-23T18:24:12.9099049Z ##[debug]loaded 9
2017-08-23T18:24:13.0879138Z ##[debug]XmlWorkingDir=C:\vsoAgent\_work\1\s
2017-08-23T18:24:13.0889663Z ##[debug]JsonPatchContent== //Parameter[@Name='dbServer']/@value => "dfg"
2017-08-23T18:24:13.0889663Z = //Parameter[@Name='testOutDir']/@value => "\tst\00_satlogs"
2017-08-23T18:24:13.0889663Z = //Parameter[@Name='applicationRootDirectory']/@value => "\Release-trunk\drop
2017-08-23T18:24:13.0899706Z ##[debug]OutputPatchFile=true
2017-08-23T18:24:13.0909256Z ##[debug]FailIfNoPatchApplied=true
2017-08-23T18:24:13.0909256Z ##[debug]SkipErrors=false
2017-08-23T18:24:13.0919119Z ##[debug]SyntaxType=slick
2017-08-23T18:24:13.0919119Z ##[debug]XmlTargetFilters=*.runsettings
2017-08-23T18:24:13.0929234Z ##[debug]Namespaces=null
2017-08-23T18:24:13.0959144Z ##[debug]task result: Failed
2017-08-23T18:24:13.0969127Z SyntaxError: Unexpected number
2017-08-23T18:24:13.1039139Z ##[error]SyntaxError: Unexpected number

**Comment by sandorfr** [comment]
Can you share the input xml?

**Comment by zaengi** [comment]
it's a normal VS2015 *.runsettings file, excerpt given here:

[val-test.runsettings.txt](https://github.com/geeklearningio/gl-vsts-tasks-file-patch/files/1248250/val-test.runsettings.txt)

**Comment by sandorfr** [comment]
There's something weird about the last line of your patch in your logs, it's missing a double quote:  `//Parameter[@name='applicationRootDirectory']/@value => "\Release-trunk\drop`
Is it coming from copy pasting the logs, if not it might be the problem.

We plan on improving those kind of error detection as Microsoft released so validation api for tasks parameters 🎉

**Comment by zaengi** [comment]
Thanks for that good input, it was wrong but didnt solve the problem:

2017-08-31T14:11:39.1871172Z = //Parameter[@Name='testOutDir']/@value => "\tst\00_satlogs"
2017-08-31T14:11:39.1881160Z = //Parameter[@Name='applicationRootDirectory']/@value => "H:\a\PDS_Release-trunk\drop"
2017-08-31T14:11:39.1921147Z ##[error]SyntaxError: Unexpected number

How can I use "Standard Syntax" instead of "Slick" with XML files?

**Comment by sandorfr** [comment]
Found the cause it's the same as #4, you have backslashes in your values so they have to be escaped.  

```
= //Parameter[@name='testOutDir']/@value => "\\tst\\00_satlogs"
= //Parameter[@name='applicationRootDirectory']/@value => "H:\\a\\PDS_Release-trunk\\drop"
```

Regarding the standard Syntax, I woudn't recommend it as it's very verbose. However if you really want to use it, you have samples in the tests : https://github.com/geeklearningio/gl-vsts-tasks-file-patch/blob/master/Tests/XmlPatch/xmlPatcher.spec.ts#L100

**Comment by zaengi** [comment]
Thanks - that solved the problem!
Only one trouble is left despite:
I want to use a common variable for the local paths, but how can I do that, when your build step needs escaped syntax and others dont?

**Comment by sandorfr** [comment]
yes you are running into the same issue as #4 for that as well and I have to respawn a discussion with vsts team as I really feel uncomfortable introducing a custom syntax to handle escaping.

Meanwhile you can use https://marketplace.visualstudio.com/items?itemName=jessehouwing.jessehouwing-vsts-variable-tasks to transform your variable to an escaped version.
